### DISCUSSION TITLE
Program successfully flashes but isn't running on the Microcontroller

### DISCUSSION THREAD

**Discussion by complexmuhm** [modm-io/modm#958] [Q&A] [answered]
I have built and flashed the example blue pill blink program on my stm32f103c8t6 and the LED won't blink. I have made sure that the microcontroller isn't dead by manually compiling, linking and flashing my own program through gnu arm. And it works. I can not get the modm blink example to work. I have tried this with a fake blue pill too. Same result: it successfully gets flashed but the code is not running.

I am using the STLINK V2 USB Programmer through SWD.v

![grafik](https://user-images.githubusercontent.com/48455449/217640614-563351e9-2ca5-49a9-9dcb-29b41a216b19.png)

**Comment by salkinium** [comment]
Does it start working if you power cycle the blue pill board without the SWD debugger attached?

**Reply by complexmuhm** [reply]
No nothing.

**Reply by rleh** [reply]
Sound to me like a BOOT0/BOOT1 pin problem:

![image](https://user-images.githubusercontent.com/2820734/217643879-1c659609-cc5e-4036-9efb-8e8e6931495b.png)

(from [STM32F103 Reference Manual RM0008](https://www.st.com/resource/en/reference_manual/cd00171190-stm32f101xx-stm32f102xx-stm32f103xx-stm32f105xx-and-stm32f107xx-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf))

**Reply by rleh** [reply]
OpenOCD writes your program to the _main flash memory_, so you have to connect the BOOT0 pin to GND (logic 0).

**Reply by complexmuhm** [reply]
I have already accounted for this. I am currently in Boot mode "Main Flash memory".

**Reply by complexmuhm** [reply]
As I said: If I compile my code only using CMSIS library with gnu arm embedded toolchain and flash it, it works. But using modm and "scons program" the compiled blink program suddenly won't work.

**Comment by rleh** [comment]
modm's blue-pill blink example [toggles](https://github.com/modm-io/modm/blob/develop/examples/blue_pill_f103/blink/main.cpp#L31-L35) gpio pin [PC13](https://github.com/modm-io/modm/blob/develop/src/modm/board/blue_pill_f103/board.hpp#L108), are you toggling the same pin in your other program?
There might be "Blue-Pill" boards with the LED connected to another pin...

Do you have any other dev board (which is [supported by modm](https://modm.io/#boards)) available to test another blink example?

**Reply by complexmuhm** [reply]
I have used the provided example and also manually setup the Gpio C13 pin like so:
![grafik](https://user-images.githubusercontent.com/48455449/217649728-43f62931-8981-4a4e-9d0c-44795ad242be.png)
But it still doesn't work. This is the other program with which it works:
![grafik](https://user-images.githubusercontent.com/48455449/217649749-759044df-56db-4797-a4a4-a525434e79fa.png)

**Reply by salkinium** [reply]
I can think of two more things: 
1. `Board::initialize()` enables the external crystal to use for clock. It may be broken, you can test this by commenting out the line in the modm example.
2. What toolchain are you using? Are you using the [toolchain from ARM](https://developer.arm.com/Tools%20and%20Software/GNU%20Toolchain), or the compiler from your distribution?

**Reply by salkinium** [reply]
Oh, and you can `scons debug` and single step through the code to find what actually happens. In GDB you can type `restart` to reset to the reset handler, then `si`, `ni` to single step around.

**Reply by complexmuhm** [reply]
thanks for the suggestion. I tried to manually step through it and also removing the `Board::initialize()`. Here is a screenshot:
![grafik](https://user-images.githubusercontent.com/48455449/217832534-f3634dd6-1327-4827-9095-b0b39ee386fc.png)
While stepping and waiting after `Led::set` and `Led::reset` I couldn't see the LED blink. The program is being run, according to the debugging. I am not sure if it is addressing wrong addresses... (*as in addressing the wrong gpio addresses)

**Reply by complexmuhm** [reply]
And it also looks like, that the program gets stuck in `modm::delay_us`
![grafik](https://user-images.githubusercontent.com/48455449/217833834-3f44aef7-f99a-470c-adfb-c309bfe65ffd.png)

**Reply by complexmuhm** [reply]
>     2. What toolchain are you using? Are you using the [toolchain from ARM](https://developer.arm.com/Tools%20and%20Software/GNU%20Toolchain), or the compiler from your distribution?
I am using the ARM GNU Toolchain, I have setup a basic project using the minimal example folder and have modified the linker script, the startup asm file and some linker flags. Here is the console output for compiling and flashing:
![grafik](https://user-images.githubusercontent.com/48455449/217834522-8214ba51-2117-401f-b6f1-e833e96ae1f0.png)

**Comment by complexmuhm** [comment, marked as answer]
Okay update: instead of using the blue pill example folder I have created my own folder and setup the project myself. It seems that it is working now.

**Reply by salkinium** [reply]
Hm, that's interesting and weird. But glad it's now working for you. If you find out why that happens (perhaps a path issue?), let us know.
I also cannot reproduce this behavior with the three blue pills that I have lying around.

**Reply by complexmuhm** [reply]
Definitely weird. I have no clue why this is happening. This is perhaps the only difference between those two projects the project.xml (left the one not working, right the one working):
![grafik](https://user-images.githubusercontent.com/48455449/217919317-e4a750ca-7610-4b69-aba8-d917720ccb16.png)

**Reply by rleh** [reply]
I also just checked the example on one of my Blue-Pill boards and it worked perfectly.
